### PR TITLE
Implementing reading Data Links

### LINKED ISSUE #9123: Initial implementation of Data Links
- [x] Introduce a datalink format that is detected based on extension `.datalink`
  - It is parsed as JSON.
  - Extend the [dataLinkSchema.json](https://github.com/enso-org/enso/blob/develop/app/ide-desktop/lib/dashboard/src/data/dataLinkSchema.json) with a version number.
	  - [x] We are also considering to extend it with the library name (e.g. `Standard.AWS`). This way if the library import is missing and the format is not recognized, we can give the user a clearer error and even provide auto-fixes in the future.
  - ~~Parser should check the version number and reject datalinks created with newer Enso version as they may contain unsupported parameters.~~ TODO: to revise later
- [x] Introduce a `DataLinkSPI` (or similar) that allows to register new types of datalinks by external libraries.
  - If a datalink type is unrecognized, report an error that the user should import a library.
  - TODO: should we tell which lib so that we could automate the import in the future?
- Implement datalinks that allow to read:
	- [x] an `S3_File`
		- [x] it should allow setting the AWS credentials and allow the keys to contain `Enso_Secret`
		- [x] it should read the underlying file and parse it using the default format
	- [x] a HTTP request

### PR DESCRIPTION
**Pull request by radeusgd**
### Pull Request Description

- Close #9123 

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

File: app/ide-desktop/lib/dashboard/src/data/dataLinkSchema.json [hunk 1]
```diff
@@ -88,32 +89,35 @@
       "type": "object",
       "properties": {
         "type": { "title": "Type", "const": "S3", "type": "string" },
+        "libraryName": { "const": "Standard.AWS" },
         "uri": {
           "title": "URI",
           "description": "Must start with \"s3://\".",
           "type": "string",
-          "pattern": "^s3://[\\w.~-]+/[/\\w.~-]+$"
+          "pattern": "^s3://[a-z0-9.-]{3,63}/.{1,1024}$"
```
Comment (radeusgd) [inline app/ide-desktop/lib/dashboard/src/data/dataLinkSchema.json:97]: The S3 bucket name can only contain _lowercase letters_, digits, `.` and `-`. It has to be at least 3 characters long and at most 63 characters long. 

There are a few more restrictions - e.g. it cannot start with `-`, it cannot contain double `--` etc. But I thought that just checking the simple ones is enough - it's just a basic sanity check - even if the bucket name is valid, the bucket could not exist / be not accessible anyway, so this does not need to be comprehensive.

The bucket key can be arbitrary, but it has a limit of 1024 bytes, so the length limit is a heuristic of that.

**Review by somebody1234 — APPROVED** [review]
approving `CODEOWNERS` and changes to the data file in the dashboard codebase.

**Comment by somebody1234** [comment]
@radeusgd should the test file not also be owned by the libraries team?

**Comment by radeusgd** [comment]
> @radeusgd should the test file not also be owned by the libraries team?

@somebody1234 I was not sure, since it's written in TypeScript I thought it may make sense for changes to be approved by someone from GUI. But I guess it _could_ be either way.

File: distribution/lib/Standard/AWS/0.0.0-dev/src/S3/S3_Data_Link.enso [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+    ## PRIVATE
+    as_file self -> S3_File = S3_File.new self.uri self.credentials
+
+    ## PRIVATE
+    read self (on_problems : Problem_Behavior) =
+        self.as_file.read self.format on_problems
+
+## PRIVATE
+parse_aws_auth json -> AWS_Credential | Nothing =
```
Comment (jdunkerley) [inline distribution/lib/Standard/AWS/0.0.0-dev/src/S3/S3_Data_Link.enso]: possibly belongs in `AWS_Credential` but we can move when we have more data links for AWS.

**Review by jdunkerley — APPROVED** [review]
Looks a great place to build on top of.

**Comment by jdunkerley** [comment]
> > @radeusgd should the test file not also be owned by the libraries team?
> 
> @somebody1234 I was not sure, since it's written in TypeScript I thought it may make sense for changes to be approved by someone from GUI. But I guess it _could_ be either way.

I agree lets add it to libs to be able to approve too.

File: distribution/lib/Standard/AWS/0.0.0-dev/src/S3/S3_Data_Link.enso (hunk 2 above)
Comment (radeusgd) [reply distribution/lib/Standard/AWS/0.0.0-dev/src/S3/S3_Data_Link.enso]: Moved to `Internal.Data_Link_Helpers`

File: test/AWS_Tests/src/S3_Spec.enso [hunk 3]
```diff
@@ -544,6 +544,28 @@ add_specs suite_builder =
                 destination_file.read . should_equal "FooBar"
                 source_file.exists . should_be_false
 
+    suite_builder.group "DataLinks to S3_File" pending=api_pending group_builder->
+        group_builder.specify "should be able to read a data link of an S3 File" <|
+            # It reads the datalink description and then reads the actual S3 file contents:
+            (enso_project.data / "simple.datalink") . read . should_equal "Hello WORLD!"
+
+        #group_builder.specify "should be able to read a data link with custom credentials and secrets" pending=cloud_setup.pending <| cloud_setup.with_prepared_environment <|
+        group_builder.specify "should be able to read a data link with custom credentials and secrets" pending="TODO: reading secrets from path" <| cloud_setup.with_prepared_environment <|
```
Comment (radeusgd) [inline test/AWS_Tests/src/S3_Spec.enso:553]: This test is enabled in the [next PR](https://github.com/enso-org/enso/pull/9225).

**Review by GregoryTravis — APPROVED** [review]

**Pull request merged**